src/nvim/os/env.c

```diff
@@ -423,43 +423,41 @@ static char *remove_tail(char *p, char *pend, char *name)
 /// @param[out] mustfree Ouput parameter for the caller to determine if they are
 ///           responsible for releasing memory. Must be initialized to false
 ///           by the caller.
-char_u *vim_getenv(char_u *name, bool *mustfree)
+char *vim_getenv(const char *name, bool *mustfree)
 {
-  char_u      *p;
-  char_u      *pend;
-  int vimruntime;
-
-
-  p = (char_u *)os_getenv((char *)name);
-  if (p != NULL && *p == NUL)       /* empty is the same as not set */
-    p = NULL;
-
-  if (p != NULL) {
-    return p;
+  const char *kos_env_path = os_getenv(name);
+  if (kos_env_path != NULL
+      && *kos_env_path == NUL) /* empty is the same as not set */
+    kos_env_path = NULL;
+
+  if (kos_env_path != NULL) {
+    *mustfree = true;
+    return xstrdup(kos_env_path);
```

Comment:
The behaviour you mention was already there before.
So, in that sense, I'm ok with these changes as they don't break anything it wasn't already broken.

Now, you're right in the sense that this function is still improvable, as it can still return things as `char *` that shouldn't be modified.

I didn't mention that because I didn't want to prolong this PR for more time.
But it probably was the wrong choice. 

It would definitely be better to always return an allocated string and remove mustfree parameter. I don't think copying a few strings can harm performance, and we'd have better code.
So, @mbainter, could you please have a look on that?